### DISCUSSION TITLE
Странное поведение linkToPanel

### DISCUSSION THREAD

**Discussion by AaaronWalsh** [totumonline/totum-ru-issues-and-discussions#517] [Как это сделать?] [answered]
Версия тотум 4.11.57.0 mit

В таблице есть столбец с кнопкой. В кнопке простой код:
= : while(action: $insert; action: $edit; action: $info)
insert: insert(в таблицу2)
edit: linkToPanel(table: таблицу2; id: $#rid; fields: $flist; refresh: true; columns: 2)
info: linkToText
Получается что функция linkToPanel не ждет закрытия окна по кнопке сохранить/отмена и сразу идет вызов info, побочным эффектом от этого получается то что refresh: true для исходной таблицы выполняется до закрытия панели

Как заставить linkToPanel  ждать закрытия?

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Это не является ошибкой — это стандартное поведение. Код действия не встает на паузу при открытии linkToPanel в вашем примере. 

Если вы хотите последовательно, то используйте linkToDataTable и в нем уже по Кнопке делаете insert с откорректированными данными и linkToDataText.

Временная таблица в этом случае ставится как скрытая из Дерева — используется как техническое окно.

Если один параметр меняется, то посмотрите в сторону linkToInput(Select)

**Reply by AaaronWalsh** [reply]
Спасибо
Можно добавить в wish-list может отдельную функцию или дополнительный параметр для linkToPanel, очень удобная функция но без ожидания сценарии ее использования сильно уменьшаются и на вставку с доп данными или редактирование нужно городить лишнего кода и визуально даже внутри окна таблицы выходит еще окно с linkToDataTable дальше еще одно с linkToPanel

**Reply by ttm-alexeypolunin** [reply]
Вот прям так там из логики работы тотума не может быть ожидания. В вашем случае linkToPanel это добавление новой строки или редактирование существующей?

Те вы делаете insert и эту строку открываете?

В таблицу в которую вы делаете insert идет запись другими способами?

**Reply by AaaronWalsh** [reply]
И тот и тот вариант. 
Таблица1 при нажатии на кнопку добавляет строку(если ее нет)  через insert в таблицу 2 и открывает окно редактирования linkToPanel этой строки. После чего приходится нажимать на вторую кнопку чтобы в таблице1 пере-считалась строка которая зависит от значений в таблице 2

**Reply by ttm-alexeypolunin** [reply]
Открывается linkToPanel, он изменяет строку, пользователь сохраняет ltp и это означает, что строка в Т2 изменилась. Вам нужно поставить отслеживание изменения строки в T2 и там сделать код действия на пересчет строки в Т1.

Делается через скрытое поле тип Данные в Т2, в котором через listCreate собираете значения отслеживаемых полей в строке. На это поле вешаете код действия на recalculate

**Reply by AaaronWalsh** [reply]
Обходные методы мне понятны, дело в том что они создают непрозрачность и в целом усложняют разработку.
А так было бы всё предельно просто и наглядно. Может добавить postaction в linkToPanel или еще что то

**Reply by ttm-alexeypolunin** [reply]
Это не обходные методы, а основные. Postaction в linkToPanel можно рассмотреть, но не рассчитывайте на него в этом году.

**Reply by ttm-alexeypolunin** [reply]
* под Основным я имею в виду то, что описанный мной подход через отслеживание строки и рекалькулейт предпочтителен, тк является независимым от типа изменения строки, через панель/таблицу/кодом действия итп.

В случае, когда вы добавляете на каждый входящий метод свое действие быстро приводит к тому, что где-то оно забывается и это всегда очень сложно отследить на отладке.
